src/util/strencodings.h

```diff
@@ -67,6 +68,66 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
 {
     return TryParseHex<Byte>(hex_str).value_or(std::vector<Byte>{});
 }
+
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xA;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+struct ConstevalHexLiteral {
+    const std::string_view inner;
+    template <size_t N>
+    consteval ConstevalHexLiteral(const char (&hex_str)[N]) : inner{hex_str} { Validate(inner); }
+    consteval ConstevalHexLiteral(const std::string_view hex_str) : inner{hex_str} { Validate(inner); }
+    consteval ConstevalHexLiteral(std::nullptr_t) = delete;
+
+    static consteval void Validate(const std::string_view hex_str)
+    {
+        std::ranges::for_each(hex_str, ConstevalHexDigit);
+
+        if (hex_str.size() % 2 == 1) throw "2 hex digits required per byte";
+    }
+};
+
+/** Validates input at compile time, but allocates the vector at runtime in
+ *  order to cross the compile time/runtime dynamic memory barrier.
+ */
+template <typename Byte = uint8_t>
+std::vector<Byte> VecFromHex(ConstevalHexLiteral hex_str)
+{
+    // No need to check HexDigit() return values or that hex_str has even length
+    // since ConstevalHexLiteral has already validated.
+    std::vector<Byte> rv(hex_str.inner.size() / 2);
+    auto it = hex_str.inner.begin();
+    for (auto& elem : rv) {
+        auto hi = HexDigit(*(it++)) << 4;
+        elem = Byte(hi | HexDigit(*(it++)));
+    }
+    return rv;
+}
+
+/** Converts from hex string literal to byte array at compile time.
+ * @warning Unlike VecFromHex which returns a vector, the returned array may use
+ *          significant stack space if called inside a function.
+ */
+template <typename Byte = uint8_t, size_t Size>
+consteval std::array<Byte, Size / 2> ArrayFromHex(const char (&hex_str)[Size])
+{
+    // Size should be uneven due to implicit null terminator.
+    if (Size % 2 == 0) throw "2 hex digits required per byte";
```

Comment:
In commit "refactor: Add consteval ArrayFromHex and VecFromHex" (b997d266e549ad6745d01a696cd9fbf1d3525944)

It seems like this is assuming the last character in the string is \0 and then ignoring it? I guess that is good because it makes this convenient to call, but in theory it means somebody could call this with a character array that is does not end in \0 and then the last character of the string would be ignored.

Would suggest changing this to:

```c++
size_t size{hex_str[Size-1] == '\0' ? Size-1 : Size};
if (size % 2 != 0) throw "2 hex digits required per byte";
```

to work correctly in either case and actually verify that the last character is \0 if an odd-size array is passed.

Reply:
Do we really need to support both?

Reply:
Continued in https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1718988155

Resolving this one for now.